Tutorial page: Using DRA and installing drivers (#51716)
kubernetes/website: content/en/docs/tutorials/cluster-management/install-use-dra.md

Reviewer: Can I just make you aware of the [highlight](https://gohugo.io/shortcodes/highlight/) shortcode @lauralorenz? It can be really useful when there is one really important line in a big configuration file / manifest / etc.

Reviewer: Combine on the same page yes; combine for the same snippet no.

Reviewer: Ideally we also update https://k8s.io/docs/tutorials/ to link to the new tutorial. But I would much sooner see this merge than worry about every last detail.

Author: Fixed in follow on PR

---

> title: Install and Use DRA Drivers

Author: Not necessarily in love with this title so open to ideas 😅

content_type: tutorial
weight: 60
min-kubernetes-server-version: v1.32
---

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

<!-- overview -->
This page then explains what to expect when installing DRA drivers in your
cluster and how to use them in conjunction with the DRA APIs to request and
observe the allocation of a Pod's hardware claim. This page is intended for
cluster administrators.

<!-- objectives -->

### {{% heading "objectives" %}}

> * Install an example DRA driver

Author: also added "and deploy" fwiw

* Deploy a Pod requesting a hardware claim using DRA APIs
* Remove the Pod and observe the cleanup

<!-- prerequisites -->
## {{% heading "prerequisites" %}}

> {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Author: The example driver has been tested with kind on linux and mac, I happen to have tested it on GKE, but not specifically on Windows or with any other node variations. Is that required and/or what is the best way to indicate those testing limitations?

> {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Reviewer: Try "Your clusters should have at least two nodes that run Linux, and that are not reserved for running the control plane".

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Your cluster also must be configured to use the Dynamic Resource Allocation

> feature. DRA APIs are stable as of Kubernetes version 1.34 but can be disabled.

Author: TODO: not sure exactly where stable vs on by default landed in 1.34, update this line.

Also, should I just copy over the #enable-dra directions from the other page and put them inline? The other page goes on to do a lot of other stuff I don't really want the tutorial user to do, but also don't want to copy pasta too much.

If your cluster has disabled them or you are using an older version of
Kubernetes, you must enable the DRA API groups as described in [Setting up
Dynamic Resource
Allocation](/docs/tasks/configure-pod-container/assign-resources/set-up-dra-cluster/#enable-dra).

<!-- lessoncontent -->

## Background

Dynamic Resource Allocation (DRA) is a Kubernetes feature that allows a cluster
to manage availability and allocation of hardware resources to satisfy Pod-based
claims for hardware requirements and preferences. To support this, a mixture of
Kubernetes native components (like the Kubernetes scheduler, kubelet, and
kube-controller-manager) and third-party components (called DRA drivers) share
the responsibility to advertise, allocate, prepare, mount, healthcheck,
unprepare, and cleanup resources throughout the Pod lifecycle. These components
share information via a series of DRA specific APIs in the

> `resource.k8s.io/v1beta2` API group, including {{< glossary_tooltip

Author: I tested all of these on kind version 1.33, as my local build of 1.34 was not behaving (yet). Once code freeze gets a lil cooler (😅 ) I will retest on 1.34, and update these references and the examples with the v1 APIs.

Reviewer: This PR should target _main_, I think, and then we want a **follow up** PR against _dev-1.34_ that updates the tutorial.

We should test the first iteration with v1.33 as well.

likely deployed as a DaemonSet to all or a selection of the nodes (using node
selectors or similar mechanisms) in your cluster.

Check your driver's documentation for specific installation instructions, which
may include a helm chart, a set of manifests, or other deployment tooling.

To illustrate a common installation procedure in this tutorial, you will use an
example driver which can be found in the
[kubernetes-sigs/dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver)
repository.

### {{% heading "procedure" %}}

#### Download the source code for the example driver

1. Clone the repository:

```shell
git clone https://github.com/kubernetes-sigs/dra-example-driver.git
cd dra-example-driver
```

#### Enable access to your DRA driver's image

In a production environment, you would likely be using a previously released or
qualified image from the driver vendor or your own organization, and your nodes
would need to have access to the image registry where the driver image is
hosted. In this tutorial, you will build an image locally to simulate access to a
DRA driver image. 

2. Build an image containing the binary for the example driver:

```shell

> ./demo/build-driver.sh

Author: Since it's already released in registry.k8s.io, and is suggested this way in comments elsewhere, I'm going with (1).

```shell
./demo/build-driver.sh
```

#### Deploy the DRA driver components

For this tutorial, you will install the critical example resource driver
components individually with `kubectl`.

{{< note >}}
You can deploy all components from this section at once from one file with

> `kubectl apply -f https://k8s.io/examples/dra/driver-install/all.yaml`.

Author: I have anything namespaced going into the default namespace for ease of use in the tutorial, but is that not recommended? Can change, just lmk.

---
title: Dynamic Resource Allocation Tutorial
content_type: tutorial
weight: 60
min-kubernetes-server-version: v1.32
---

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

<!-- overview -->
This tutorial explains what to expect when installing
{{< glossary_tooltip term_id="dra" text="DRA" >}} drivers in your
cluster and how to use them in conjunction with the DRA APIs to request and
observe the allocation of a Pod's hardware claim. This page is intended for

> cluster administrators.

Author: I agree that it is useful to workload authors, but I see the primary audience of this tutorial as cluster administrators, who I feel also need to understand how the DRA APIs interact with the driver to maintain them well. If I had to choose which audience to focus on in this doc, I would choose cluster administrators. If the prose gets too muddied by considering both of them, I'd rather make a second tutorial for the workload authors that doesn't go into the driver at all (since IMHO that is more representative of what we think their experience should be like) and also potentially goes into much more detailed/ ✨ fun ✨  use cases.

> cluster administrators.

Author: x-post from other comment too: I've added this rationale in an HTML comment for future maintainers in https://github.com/kubernetes/website/pull/51716/commits/6bf1164f42488d01c5142b70e68f854fafcdf571

## {{% heading "prerequisites" %}}

Your cluster should support [RBAC](/docs/reference/access-authn-authz/rbac/).
You can try this tutorial with a cluster using
a different authorization mechanism, but in that case you will have to adapt the
steps around defining roles and permissions.

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Your cluster also must be configured to use the Dynamic Resource Allocation
feature.
To enable the DRA feature, you must enable the following feature gates and API groups:

1.  Enable the `DynamicResourceAllocation`
    [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
    on all of the following components:
   
    * `kube-apiserver`
    * `kube-controller-manager`
    * `kube-scheduler`
    * `kubelet`

1.  Enable the following
    {{< glossary_tooltip text="API groups" term_id="api-group" >}}:

    * `resource.k8s.io/v1beta1`: required for DRA to function.
    * `resource.k8s.io/v1beta2`: optional, recommended improvements to the user
      experience.
     
    For more information, see
    [Enabling or disabling API groups](/docs/reference/using-api/#enabling-or-disabling).

<!-- lessoncontent -->

> ## {{% heading "synopsis" %}}  

Reviewer: Yes, I agree.

> Dynamic Resource Allocation (DRA) is a Kubernetes feature that allows a cluster

Reviewer: Yes, or use the shortcode to transclude the glossary definition, or both.

to manage availability and allocation of hardware resources to satisfy Pod-based
claims for hardware requirements and preferences. To support this, a mixture of
Kubernetes native components (like the Kubernetes scheduler, kubelet, and
kube-controller-manager) and third-party components (called DRA drivers) share
the responsibility to advertise, allocate, prepare, mount, healthcheck,
unprepare, and cleanup resources throughout the Pod lifecycle. These components
share information via a series of DRA specific APIs in the
`resource.k8s.io/v1beta2` API group, including {{< glossary_tooltip
text="DeviceClasses" term_id="deviceclass" >}}, ResourceSlice, {{<
glossary_tooltip text="ResourceClaims" term_id="resourceclaim" >}}, as well as
new fields in the Pod spec itself.

## Explore the DRA initial state

With no driver installed or Pod claims yet to satisfy, you can observe the
initial state of a cluster with DRA enabled.

### Procedure

> ### Check the DeviceClasses in your cluster

Author: I've updated this to follow the suggested format. I will update the concepts page with anything I feel is missing there from the prose that was once in this section, since I felt it had some good wisdom gems in there and good review done to it, but in a separate PR.

Note to self: stuff to carry over to concepts page if warranted is in https://github.com/kubernetes/website/pull/51716/commits/a5633a1668363f8f4a9ef2bb047a9b333f3c056e

The {{< api-reference page="extend-resources/device-class-v1beta2" >}} resources represent a centralized list of the device
classes known to the cluster, each managed by a uniquely named
DRA driver. If you set up a new test cluster for this tutorial, there should be no
DeviceClasses.

1. Check with `kubectl`:
```shell
kubectl get deviceclasses
```
The output is similar to this:
```
No resources found
```

### Check on ResourceSlices
A ResourceSlice is a partial list of the [{< glossary_tooltip text="infrastructure resources" term_id="infrastructure-resource" >}} that are potentially available to use from Nodes. It is a partial list. Some infrastructure resource types (such as CPU and memory)
don't need to be claimed, and are handled through other mechanisms.
Storage (as in files and block devices) has its own management mechanism too; see [Storage](/docs/concepts/storage/volumes) elsewhere in the documentation.

ResourceSlices can represent existing allocated infrastructure, but they can also

> represent an offer to provide infrastructure. For example, a specialized driver

Author: Resolving because this prose itself has been removed from here in favor of linking out to the concepts page, but see https://github.com/kubernetes/website/pull/51716#discussion_r2268178915 for more on that and the intent to follow up on the concepts page to make sure these caveats are present there.

kubectl get resourceclaimtemplates -A
```
The output is similar to this:
```
No resources found
No resources found
```

### Section summary  

At this point, you have confirmed that DRA is enabled and configured properly in
the cluster, and that no DRA drivers have advertised any resources to the DRA
APIs yet.

## Install an example DRA driver {#install-example-driver}  

DRA drivers are third-party applications that run on each node of your cluster
to interface with the hardware of that node and Kubernetes' native DRA
components. The installation procedure depends on the driver you choose, but is
likely deployed as a {{< glossary_tooltip term_id="daemonset" >}} to all or a selection of the nodes (using {{< glossary_tooltip text="selectors" term_id="selector" >}} or similar mechanisms) in your cluster.

Check your driver's documentation for specific installation instructions, which
may include a helm chart, a set of manifests, or other deployment tooling.

To illustrate a common installation procedure in this tutorial, you will use an
example driver which can be found in the
[kubernetes-sigs/dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver)
repository.

> ### Procedure

Reviewer: ```suggestion
```

### Create a namespace for the tutorial

To make it easier to cleanup later, create a namespace called `dra-tutorial` in your cluster.

```shell
kubectl create namespace dra-tutorial 
```

### Enable access to your DRA driver's image

In a production environment, you would likely be using a previously released or
qualified image from the driver vendor or your own organization, and your nodes
would need to have access to the image registry where the driver image is
hosted. In this tutorial, you will use a public released image of the dra-example-driver to simulate access to a
DRA driver image. You can confirm your nodes have access to it by running the following from within one of your cluster's nodes:

2. Test pull the image for the example DRA driver.

```shell
docker pull registry.k8s.io/dra-example-driver/dra-example-driver:v0.1.0
```

### Deploy the DRA driver components

For this tutorial, you will install the critical example resource driver

> components individually with `kubectl`.

Author: To be clear, from my POV the original point of this doc is not "just give me something I can use to play with DRA", it's for people (mainly cluster administrators) to understand the importance of the driver and its interaction with the DRA APIs. That is where the requirement to not use helm and to be more direct with all the installation procedures came from. That being said I'm open to including helm in a second tab where relevant.

> components individually with `kubectl`.

Author: I've added this rationale in an HTML comment for future maintainers in https://github.com/kubernetes/website/pull/51716/commits/6bf1164f42488d01c5142b70e68f854fafcdf571

```
The output is similar to this:
```
NAME                                     READY   STATUS    RESTARTS   AGE
dra-example-driver-kubeletplugin-4sk2x   1/1     Running   0          13s
dra-example-driver-kubeletplugin-cttr2   1/1     Running   0          13s
```

### Explore the DRA state

The initial reponsibility of each node's local DRA driver is to update the
cluster with what classes of devices are available to Pods on that node, by publishing its
metadata to the ResourceSlices API. You can check that API to see that each node
with a driver is advertising the device class it represents.

7. Check for available ResourceSlices (using `kubectl`): 

```shell
kubectl get resourceslices
```
The output is similar to this:
```
NAME                                 NODE           DRIVER            POOL           AGE
kind-worker-gpu.example.com-k69gd    kind-worker    gpu.example.com   kind-worker    19s
kind-worker2-gpu.example.com-qdgpn   kind-worker2   gpu.example.com   kind-worker2   19s
```

### Section summary

At this point, you have successfully installed the example DRA driver, and
confirmed its initial configuration. You're now ready to use DRA to schedule
Pods.

> ## Deploy a Pod with a claim using DRA APIs

Reviewer 2: ```suggestion
## Claim resources and deploy a Pod
```

> Pods can make claims on what type of hardware they require and/or prefer against
> an instance of the DRA ResourceClaim API. In the example driver, a
> `capacity.memory` attribute is exposed for mock GPU devices. You will leverage
> that attribute using {{< glossary_tooltip term_id="cel" >}} to express a highly
> customizable requirement in a ResourceClaim, then attach that ResourceClaim to a
> Pod via its manifest and observe the deployment.

Reviewer 2: ```suggestion
To request resources using DRA, you create ResourceClaims or ResourceClaimTemplates that define the resources that your Pods need.
In the example driver, a `capacity.memory` attribute is exposed for mock GPU devices. This section shows you how to use {{< glossary_tooltip term_id="cel" >}} to  express your requirements in a ResourceClaim, select that ResourceClaim in a Pod specification, and observe the resource allocation.
```

This tutorial showcases only one basic example of a DRA  
ResourceClaim. Read [Dynamic Resource  
Allocation](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)  
to learn more about ResourceClaims.  

### Procedure

### Create the ResourceClaim

The Pod manifest itself will need to include a reference to a ResourceClaim, so
you will make the ResourceClaim first. This is one of the simpler ResourceClaims

> possible. The request itself is composed of a {{< glossary_tooltip term_id="cel" >}} expression that will accept

Author: See https://github.com/kubernetes/website/pull/51716/commits/f2143f0a1f67903734abfd23bacde1388c70d3f9 and rendered direct fragment link https://deploy-preview-51716--kubernetes-io-main-staging.netlify.app/docs/tutorials/cluster-management/install-use-dra/#create-the-resourceclaim to see if this is better.

> any GPU advertising over 10Gi memory capacity by a `gpu.example.com` driver.

Author: See https://github.com/kubernetes/website/pull/51716/commits/f2143f0a1f67903734abfd23bacde1388c70d3f9 and rendered direct fragment link https://deploy-preview-51716--kubernetes-io-main-staging.netlify.app/docs/tutorials/cluster-management/install-use-dra/#create-the-resourceclaim to see if this is better.

> any GPU advertising over 10Gi memory capacity by a `gpu.example.com` driver.

Reviewer 3: Definitely better, thanks!

Sorry I didn't catch this the first time around, but next time I would avoid language like "The attribute exposing capacity" in case that could cause confusion between the `attributes` and `capacity` fields in the API. I think the context in this case is enough where it's still clear though, so definitely not a blocker.

Note that the name of the claim is set to `some-gpu`.

{{% code_sample language="yaml" file="dra/driver-install/example/resourceclaim.yaml" %}}

```shell
kubectl apply --server-side -f https://k8s.io/examples/dra/driver-install/example/resourceclaim.yaml
```

### Create the Pod that references that ResourceClaim

Below is the Pod manifest referencing the ResourceClaim you just made,
`some-gpu`, in the `spec.resourceClaims.resourceClaimName` field. The local name
for that claim, `gpu`, is then used in the
`spec.containers.resources.claims.name` field to allocate the claim to the Pod's
underlying container.

{{% code_sample language="yaml" file="dra/driver-install/example/pod.yaml" %}}

```shell
kubectl apply --server-side -f https://k8s.io/examples/dra/driver-install/example/pod.yaml
```

### Explore the DRA state

The cluster now tries to schedule that Pod to a node where Kubernetes can  

> satisfies the ResourceClaim. In our  

Reviewer: Check trailing whitespace.

      uid: fa55b59b-d28d-4f7d-9e5b-ef4c8476dff5
kind: List
metadata:
  resourceVersion: ""
```

4. Observe the driver by checking the pod logs for pods backing the driver
   daemonset:

```shell
kubectl logs -l app.kubernetes.io/name=dra-example-driver -n dra-tutorial
```

The output is similar to this:
```
I0729 05:11:52.679057       1 driver.go:84] NodePrepareResource is called: number of claims: 1
I0729 05:11:52.684450       1 driver.go:112] Returning newly prepared devices for claim '79e1e8d8-7e53-4362-aad1-eca97678339e': [&Device{RequestNames:[some-gpu],PoolName:kind-worker,DeviceName:gpu-4,CDIDeviceIDs:[k8s.gpu.example.com/gpu=common k8s.gpu.example.com/gpu=79e1e8d8-7e53-4362-aad1-eca97678339e-gpu-4],}]
```

### Section summary

You have now successfully deployed a Pod with a DRA based claim, and seen it
scheduled to an appropriate node and the associated DRA APIs updated to reflect
its status.

## Remove the Pod with a claim

When a Pod with a claim is deleted, the DRA driver deallocates the resource so
it can be available for future scheduling. You can observe that by deleting our
pod with a claim and seeing that the state of the ResourceClaim changes.

> ### Procedure

Reviewer: ```suggestion
```

###  Delete the pod using the resource claim

1. Use `kubectl` to delete the pod directly:

```shell
kubectl delete pod pod0 -n dra-tutorial
```

The output is similar to this:

```
pod "pod0" deleted
```

### Observe the DRA state

The driver will deallocate the hardware and update the corresponding
ResourceClaim resource that previously held the association.

2. Use `kubectl` to check the ResourceClaim is now pending:

```shell
kubectl get resourceclaims -n dra-tutorial
```

The output is similar to this:
```
NAME       STATE     AGE
some-gpu   pending   76s
```

3. Observe the driver logs and see that it processed unpreparing the device for
   this claim:

```shell
kubectl logs -l app.kubernetes.io/name=dra-example-driver -n dra-tutorial
```
The output is similar to this:
```
I0729 05:13:02.144623       1 driver.go:117] NodeUnPrepareResource is called: number of claims: 1
```

### Section summary

You have now deleted a Pod that had a claim, and observed that the driver took
action to unprepare the underlying hardware resource and update the DRA APIs to
reflect that the resource is available again for future scheduling.

## {{% heading "cleanup" %}}

> 

Reviewer: If we [recommend Helm](https://github.com/kubernetes/website/pull/51716#discussion_r2249184268) then let's recommend a Helm based cleanup here, and also be thorough about manual cleanup in case they mess up the Helm part.

> To cleanup the resources, delete the namespace for the tutorial which will clean up the ResourceClaims, driver components, and RBAC objects. Then also delete the cluster level DeviceClass resource.

Reviewer: Does this fully clean up RBAC?

weight: 60
min-kubernetes-server-version: v1.32
---
<!-- FUTURE MAINTAINERS: 
The original point of this doc was for people (mainly cluster administrators) to
understand the importance of the DRA driver and its interaction with the DRA
APIs. As a result it was a requirement of this tutorial to not use Helm and to
be more direct with all the component installation procedures. While much of
this content is also useful to workload authors, I see the primary audience of
_this_ tutorial as cluster administrators, who I feel also need to understand
how the DRA APIs interact with the driver to maintain them well. If I had to
choose which audience to focus on in this doc, I would choose cluster
administrators. If the prose gets too muddied by considering both of them, I'd
rather make a second tutorial for the workload authors that doesn't go into the
driver at all (since IMHO that is more representative of what we think their
experience should be like) and also potentially goes into much more detailed/ ✨
fun ✨ use cases. 
-->

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

<!-- overview -->
This tutorial shows you how to install {{< glossary_tooltip term_id="dra"

> text="DRA" >}} drivers in your cluster and how to use them in conjunction with

Author: Fixed in follow up PR

the DRA APIs to allocate {{< glossary_tooltip text="devices" term_id="device"
>}} to Pods. This page is intended for cluster administrators.

{{< glossary_tooltip text="Dynamic Resource Allocation (DRA)" term_id="dra" >}}
is a Kubernetes feature that allows a cluster to manage availability and
allocation of hardware resources to satisfy Pod-based claims for hardware
requirements and preferences (see the [DRA Concept

> page](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) for more

Author: Fixed in follow up PR

background). To support this, a mixture of Kubernetes built-in components (like
the Kubernetes scheduler, kubelet, and kube-controller-manager) and third-party

> components (called DRA drivers) share the responsibility to advertise, allocate,

Author: Fixed in follow up PR

* Delete a Pod that has a claim

<!-- prerequisites -->
## {{% heading "prerequisites" %}}

Your cluster should support [RBAC](/docs/reference/access-authn-authz/rbac/).
You can try this tutorial with a cluster using
a different authorization mechanism, but in that case you will have to adapt the
steps around defining roles and permissions.

{{< include "task-tutorial-prereqs.md" >}}

This tutorial has been tested with Linux nodes, though it may also work with
other types of nodes.

 {{< version-check >}}

Your cluster also must be configured to use the Dynamic Resource Allocation
feature.
To enable the DRA feature, you must enable the following feature gates and API groups:

1.  Enable the `DynamicResourceAllocation`
    [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
    on all of the following components:
   
    * `kube-apiserver`
    * `kube-controller-manager`
    * `kube-scheduler`
    * `kubelet`

1.  Enable the following
    {{< glossary_tooltip text="API groups" term_id="api-group" >}}:

    * `resource.k8s.io/v1beta1`: required for DRA to function.
    * `resource.k8s.io/v1beta2`: optional, recommended improvements to the user

>       experience.

Reviewer 3: Can we mention something like how v1beta2 isn't required for DRA to work in general, but it is required for the steps in the tutorial since it involves creating a ResourceClaim defined in that version? Or if that muddies this too much, only say to enable both v1beta1 and v1beta2 without offering details as to why each is required? Or redefine the ResourceClaim in v1beta1?

     
    For more information, see
    [Enabling or disabling API groups](/docs/reference/using-api/#enabling-or-disabling).

<!-- lessoncontent -->

> ## Explore the DRA initial state

Author: Fixed in follow up PR

    The output is similar to this:
    ```
    No resources found
    No resources found
    ```

    If you set up a new blank cluster for this tutorial, it's normal to find that
    there are no ResourceClaims or ResourceClaimTemplates as you, the user, have
    not created any. [Learn more about ResourceClaims and ResourceClaimTemplates
    here.](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#resourceclaims-templates)

At this point, you have confirmed that DRA is enabled and configured properly in
the cluster, and that no DRA drivers have advertised any resources to the DRA
APIs yet.

## Install an example DRA driver {#install-example-driver}  

DRA drivers are third-party applications that run on each node of your cluster
to interface with the hardware of that node and Kubernetes' built-in DRA
components. The installation procedure depends on the driver you choose, but is
likely deployed as a {{< glossary_tooltip term_id="daemonset" >}} to all or a
selection of the nodes (using {{< glossary_tooltip text="selectors"
term_id="selector" >}} or similar mechanisms) in your cluster.

Check your driver's documentation for specific installation instructions, which

> may include a Helm chart, a set of manifests, or other deployment tooling.

Author: Fixed in follow up PR

This tutorial uses an example driver which can be found in the
[kubernetes-sigs/dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver)

> repository to demonstrate driver installation.

Author: Fixed in follow up PR (though I wrote "for your Pods to interact with")

### Prepare your cluster for driver installation

> To make it easier to cleanup later, create a namespace called `dra-tutorial` in your cluster.

Author: Fixed in a follow up PR

be used by the driver to gain permissions to interact with the Kubernetes API
on this cluster:

      1.  Create the Service Account: 

          {{% code_sample language="yaml" file="dra/driver-install/serviceaccount.yaml" %}}

          ```shell
          kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/serviceaccount.yaml
          ```

      1.  Create the ClusterRole: 

          {{% code_sample language="yaml" file="dra/driver-install/clusterrole.yaml" %}}

          ```shell
          kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/clusterrole.yaml
          ```

      1.  Create the ClusterRoleBinding:

          {{% code_sample language="yaml" file="dra/driver-install/clusterrolebinding.yaml" %}}

          ```shell
          kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/clusterrolebinding.yaml
          ```

1.  Create a {{< glossary_tooltip term_id="priority-class" >}} for the DRA
    driver. The DRA driver component is responsible for important lifecycle
    operations for Pods with claims, so you don't want it to be preempted. Learn

>     more about [pod priority and preemption

Author: Fixed in a follow up PR

    here](/docs/concepts/scheduling-eviction/pod-priority-preemption/). Learn
    more about [good practices when maintaining a DRA driver
    here](/docs/concepts/cluster-administration/dra/).

    {{% code_sample language="yaml" file="dra/driver-install/priorityclass.yaml" %}}

    ```shell
    kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/priorityclass.yaml
    ```

1.  Deploy the actual DRA driver as a DaemonSet configured to run the example
   driver binary with the permissions provisioned above.

    {{% code_sample language="yaml" file="dra/driver-install/daemonset.yaml" %}}

    ```shell
    kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/daemonset.yaml
    ```
    It is configured with
      the volume mounts necessary to interact with the underlying Container Device
      Interface (CDI) directory, and to expose its socket to kubelet via the
      kubelet plugins directory.

> ### Verify the DRA driver installation

Author: Fixed in a follow up PR

    kind-worker-gpu.example.com-k69gd    kind-worker    gpu.example.com   kind-worker    19s
    kind-worker2-gpu.example.com-qdgpn   kind-worker2   gpu.example.com   kind-worker2   19s
    ```

At this point, you have successfully installed the example DRA driver, and
confirmed its initial configuration. You're now ready to use DRA to schedule
Pods.

## Claim resources and deploy a Pod  

To request resources using DRA, you create ResourceClaims or
ResourceClaimTemplates that define the resources that your Pods need. In the
example driver, a memory capacity attribute is exposed for mock GPU devices.
This section shows you how to use {{< glossary_tooltip term_id="cel" >}} to
express your requirements in a ResourceClaim, select that ResourceClaim in a Pod
specification, and observe the resource allocation.

This tutorial showcases only one basic example of a DRA ResourceClaim. Read
[Dynamic Resource
Allocation](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) to
learn more about ResourceClaims. 

### Create the ResourceClaim

The Pod manifest itself will include a reference to its relevant ResourceClaim

> object, which you will create now. Whatever the claim, the `deviceClassName` is

Author: Fixed in a follow up PR

expression that references attributes that may be advertised by the driver
managing that device class. 

In this example, you will create a request for any GPU advertising over 10Gi
memory capacity. The attribute exposing capacity from the example driver takes
the form `device.capacity['gpu.example.com'].memory`. Note also that the name of
the claim is set to `some-gpu`.

{{% code_sample language="yaml" file="dra/driver-install/example/resourceclaim.yaml" %}}

```shell
kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/example/resourceclaim.yaml
```

### Create the Pod that references that ResourceClaim

Below is the Pod manifest referencing the ResourceClaim you just made,
`some-gpu`, in the `spec.resourceClaims.resourceClaimName` field. The local name
for that claim, `gpu`, is then used in the
`spec.containers.resources.claims.name` field to allocate the claim to the Pod's
underlying container.

{{% code_sample language="yaml" file="dra/driver-install/example/pod.yaml" %}}

```shell
kubectl apply --server-side -f http://k8s.io/examples/dra/driver-install/example/pod.yaml
```

### Explore the DRA state

> The cluster now tries to schedule that Pod to a node where Kubernetes can

Reviewer 2: ```suggestion
After you create the Pod, the cluster tries to schedule that Pod to a node where Kubernetes can
```

Author: Fixed in a follow up PR

> satisfy the ResourceClaim. In our situation, the DRA driver is deployed on all

Reviewer 2: ```suggestion
satisfy the ResourceClaim. In this tutorial, the DRA driver is deployed on all
```

Author: Fixed in a follow up PR

nodes, and is advertising mock GPUs on all nodes, all of which have enough

> capacity advertised to satisfy the Pod's claim, so this Pod may be scheduled to
> any node and any of the mock GPUs on that node may be allocated.

Reviewer 2: ```suggestion
capacity advertised to satisfy the Pod's claim, so Kubernetes can schedule this Pod
on any node and can allocate any of the mock GPUs on that node.
```

Author: Fixed in a follow up PR

> The mock GPU driver injects environment variables in each container it is

Reviewer 2: ```suggestion
When Kubernetes allocates a mock GPU to a Pod, the example driver adds
environment variables to each container to indicate which GPUs _would_ have been
allocated to the container by a real resource driver and how those GPUs would have
been configured. To see how the Pods have been handled by DRA, check these
environment variables.
```

Author: Fixed in a follow up PR

allocated to in order to indicate which GPUs _would_ have been injected into
them by a real resource driver and how they would have been configured, so you
can check those environment variables to see how the Pods have been handled by
the system.

1.  Confirm the pod has deployed:

    ```shell
    kubectl get pod pod0 -n dra-tutorial
    ```

    The output is similar to this:
    ```
    NAME   READY   STATUS    RESTARTS   AGE
    pod0   1/1     Running   0          9s

>     ```

Reviewer 2: Probably better off in the previous section, to verify that those steps worked?

> 1.  Observe the pod logs which report the name of the mock GPU allocated:

Reviewer 2: ```suggestion
1.  Check the Pod logs, which report the name of the mock GPU that was allocated:
```

Author: Fixed in a follow up PR

    ```shell
    kubectl logs pod0 -c ctr0 -n dra-tutorial | grep -E "GPU_DEVICE_[0-9]+=" | grep -v "RESOURCE_CLAIM"
    ```

    The output is similar to this:
    ```
    declare -x GPU_DEVICE_4="gpu-4"
    ```

> 1.  Observe the ResourceClaim object:

Reviewer 2: ```suggestion
1.  Check the state of the ResourceClaim object:
```

Author: Fixed in a follow up PR

    You can observe the ResourceClaim more closely, first only to see its state

>     is allocated and reserved.

Reviewer 2: Move this after the output block. like "In this output, the `STATE` column shows that the ResourceClaim is allocated and reserved."

Author: Fixed in a follow up PR

    ```shell
    kubectl get resourceclaims -n dra-tutorial
    ```

    The output is similar to this:

    ```
    NAME       STATE                AGE
    some-gpu   allocated,reserved   34s
    ```

>     Looking deeper at the `some-gpu` ResourceClaim, you can see that the status stanza includes information about the
>     device that has been allocated and for what pod it has been reserved for:

Reviewer 2: ```suggestion
 1. Check the details of the `some-gpu` ResourceClaim. The `status` field of the ResourceClaim has information about the allocated device and the Pod that the device was reserved for.
```

        name: some-gpu
        namespace: dra-tutorial
        resourceVersion: "58357"
        uid: 79e1e8d8-7e53-4362-aad1-eca97678339e
      spec:
        devices:
          requests:
          - exactly:
              allocationMode: ExactCount
              count: 1
              deviceClassName: gpu.example.com
              selectors:
              - cel:
                  expression: device.capacity['gpu.example.com'].memory.compareTo(quantity('10Gi'))
                    >= 0
            name: some-gpu
      status:
        allocation:
          devices:
            results:
            - adminAccess: null
              device: gpu-4
              driver: gpu.example.com
              pool: kind-worker
              request: some-gpu
          nodeSelector:
            nodeSelectorTerms:
            - matchFields:
              - key: metadata.name
                operator: In
                values:
                - kind-worker
        reservedFor:
        - name: pod0
          resource: pods
          uid: fa55b59b-d28d-4f7d-9e5b-ef4c8476dff5
    kind: List
    metadata:
      resourceVersion: ""
    {{< /highlight >}}

> 1.  Observe the driver by checking the pod logs for pods backing the driver
>    daemonset:

Reviewer 2: ```suggestion
1.  To check how the driver handled device allocation, get the logs for the driver DaemonSet Pods:
```

    ```shell
    kubectl logs -l app.kubernetes.io/name=dra-example-driver -n dra-tutorial
    ```

    The output is similar to this:
    ```
    I0729 05:11:52.679057       1 driver.go:84] NodePrepareResource is called: number of claims: 1
    I0729 05:11:52.684450       1 driver.go:112] Returning newly prepared devices for claim '79e1e8d8-7e53-4362-aad1-eca97678339e': [&Device{RequestNames:[some-gpu],PoolName:kind-worker,DeviceName:gpu-4,CDIDeviceIDs:[k8s.gpu.example.com/gpu=common k8s.gpu.example.com/gpu=79e1e8d8-7e53-4362-aad1-eca97678339e-gpu-4],}]
    ```

> You have now successfully deployed a Pod with a DRA based claim, and seen it
> scheduled to an appropriate node and the associated DRA APIs updated to reflect
> its status.

Reviewer 2: ```suggestion
You have now successfully deployed a Pod that claimed devices using DRA, verified that
the Pod was scheduled to an appropriate node, and saw that the associated DRA API kinds were
updated with the allocation status.
```

Author: Fixed in a follow up PR

> ## Remove the Pod with a claim

Reviewer 2: ```suggestion
## Delete a Pod that has a claim {#delete-pod-claim}
```

Author: Fixed in a follow up PR

When a Pod with a claim is deleted, the DRA driver deallocates the resource so

> it can be available for future scheduling. You can observe that by deleting our
> pod with a claim and seeing that the state of the ResourceClaim changes.

Reviewer 2: ```suggestion
it can be available for future scheduling. To validate this behavior, delete the Pod that you
created in the previous steps and watch the corresponding changes to the ResourceClaim
and driver.
```

Author: Fixed in a follow up PR

> ###  Delete the pod using the resource claim

Reviewer 2: ```suggestion
```

> 1.  Delete the pod directly:

Reviewer 2: ```suggestion
1.  Delete the `pod0` Pod:
```

    ```shell
    kubectl delete pod pod0 -n dra-tutorial
    ```

    The output is similar to this:

    ```
    pod "pod0" deleted
    ```

> 
> ### Observe the DRA state
> 
> The driver will deallocate the hardware and update the corresponding
> ResourceClaim resource that previously held the association.
> 
> 1.  Check the ResourceClaim is now pending:

Reviewer 2: ```suggestion
    When the Pod is deleted, the driver deallocates the device from the ResourceClaim and updates
    the ResourceClaim resource in the Kubernetes API. The ResourceClaim has a `pending` state until
    it's referenced in a new Pod.

1.  Check the state of the `some-gpu` ResourceClaim:
```

Author: Fixed in follow up PR

    ```shell
    kubectl get resourceclaims -n dra-tutorial
    ```

    The output is similar to this:
    ```
    NAME       STATE     AGE
    some-gpu   pending   76s
    ```

> 1.  Observe the driver logs and see that it processed unpreparing the device for

Reviewer 2: ```suggestion
1.  Verify that the driver processed unpreparing the device for
```

Author: Fixed in follow up PR

   this claim:

    ```shell
    kubectl logs -l app.kubernetes.io/name=dra-example-driver -n dra-tutorial
    ```
    The output is similar to this:
    ```
    I0729 05:13:02.144623       1 driver.go:117] NodeUnPrepareResource is called: number of claims: 1
    ```

You have now deleted a Pod that had a claim, and observed that the driver took
action to unprepare the underlying hardware resource and update the DRA APIs to
reflect that the resource is available again for future scheduling.

## {{% heading "cleanup" %}}

> To cleanup the resources, delete the namespace for the tutorial which will clean up the ResourceClaims, driver components, and ServiceAccount. Then also delete the cluster level DeviceClass resource and cluster level RBAC resources.

Reviewer 2: ```suggestion
To clean up the resources that you created in this tutorial, follow these steps:

1. To delete the ResourceClaim, the DRA driver DaemonSet, and the ServiceAccount, delete the `dra-tutorial` namespace:
    <command>
 1. Delete the DeviceClass, the ClusterRole, and the ClusterRoleBinding:
```

> To cleanup the resources, delete the namespace for the tutorial which will clean up the ResourceClaims, driver components, and ServiceAccount. Then also delete the cluster level DeviceClass resource and cluster level RBAC resources.

Reviewer 2: Shouldn't the driver clean up the DeviceClass when it's deleted?

Author: In the case of the example driver, it doesn't; the DeviceClass is in fact deployed (and undeployed) as part of the helm chart under normal operation. Tagging this [here in the scratch for forthcoming driver author documentation](https://docs.google.com/document/d/11Ki7CYwhBdd5ZRlnLRrvFD0G_chTDJShsaDS6i6VNWo/edit?tab=t.0#bookmark=id.9dizbhh2957t ). Some issues with letting the driver do that are:
* A given node local driver can't know if the DeviceClass is still representing drivers functioning on other nodes
* In seamless upgrade case, the downturning driver shouldn't remove the DeviceClass underneath the upcoming driver

```shell
kubectl delete namespace dra-tutorial
kubectl delete deviceclass gpu.example.com
kubectl delete clusterrole dra-example-driver-role

> kubectl delete clusterrolebinding dra-example-driver-role-binding

Reviewer 3: ```suggestion
kubectl delete clusterrolebinding dra-example-driver-role-binding
kubectl delete priorityclass dra-driver-high-priority
```
